histogram.go

```diff
@@ -0,0 +1,259 @@
+	}
+	return fmt.Sprintf("%v", values)
+}
+
+// AsValues implements Buckets and returns float64
+// representations of the time.Duration values divided by time.Second.
+func (v DurationBuckets) AsValues() []float64 {
+	values := make([]float64, len(v))
+	for i := range values {
+		values[i] = float64(v[i]) / float64(time.Second)
+	}
+	return values
+}
+
+// AsDurations implements Buckets.
+func (v DurationBuckets) AsDurations() []time.Duration {
+	return []time.Duration(v)
+}
+
+// BucketPairs creates a set of bucket pairs from a set
+// of buckets describing the lower and upper bounds for
+// each derived bucket.
+func BucketPairs(buckets Buckets) []BucketPair {
+	if buckets == nil || buckets.Len() < 1 {
+		return []BucketPair{
+			bucketPair{
+				-math.MaxFloat64, math.MaxFloat64,
+				time.Duration(math.MinInt64), time.Duration(math.MaxInt64)},
+		}
+	}
+
+	if durationBuckets, ok := buckets.(DurationBuckets); ok {
+		// If using duration buckets separating negative times and
+		// positive times is very much desirable as depending on the
+		// reporter will create buckets "-infinity,0" and "0,{first_bucket}"
+		// instead of just "-infinity,{first_bucket}" which for time
+		// durations is not desirable nor pragmatic
+		hasZero := false
+		for _, b := range buckets.AsDurations() {
+			if b == 0 {
+				hasZero = true
+				break
+			}
+		}
+		if !hasZero {
+			buckets = append(DurationBuckets{0}, durationBuckets...)
+		}
+	}
+
+	// Sort before iterating to create pairs
+	sort.Sort(buckets)
+
+	var (
+		asValueBuckets    = buckets.AsValues()
+		asDurationBuckets = buckets.AsDurations()
+		pairs             = make([]BucketPair, 0, buckets.Len()+2)
+	)
+	pairs = append(pairs, bucketPair{
+		-math.MaxFloat64, asValueBuckets[0],
+		time.Duration(math.MinInt64), asDurationBuckets[0]})
+
+	prevValueBucket, prevDurationBucket :=
+		asValueBuckets[0], asDurationBuckets[0]
+	for i := 1; i < buckets.Len(); i++ {
+		pairs = append(pairs, bucketPair{
+			prevValueBucket, asValueBuckets[i],
+			prevDurationBucket, asDurationBuckets[i]})
+		prevValueBucket, prevDurationBucket =
+			asValueBuckets[i], asDurationBuckets[i]
+	}
+
+	pairs = append(pairs, bucketPair{
+		prevValueBucket, math.MaxFloat64,
+		prevDurationBucket, time.Duration(math.MaxInt64)})
+
+	return pairs
+}
+
+type bucketPair struct {
+	lowerBoundValue    float64
+	upperBoundValue    float64
+	lowerBoundDuration time.Duration
+	upperBoundDuration time.Duration
+}
+
+func (p bucketPair) LowerBoundValue() float64 {
+	return p.lowerBoundValue
+}
+
+func (p bucketPair) UpperBoundValue() float64 {
+	return p.upperBoundValue
+}
+
+func (p bucketPair) LowerBoundDuration() time.Duration {
+	return p.lowerBoundDuration
+}
+
+func (p bucketPair) UpperBoundDuration() time.Duration {
+	return p.upperBoundDuration
+}
+
+// LinearValueBuckets creates a set of linear value buckets.
+func LinearValueBuckets(start, width float64, count int) ValueBuckets {
+	if count <= 0 {
+		panic("count needs to be > 0")
```

Comment:
Yeah I'm going to return errors then add separate methods such as `MustMakeLinearValueBuckets(...)`, etc to return a single value for fluid API use.